### PR TITLE
CT-223 - algorand build tx

### PR DESCRIPTION
**Pull request by twtaylorbitgo**
Enabling algorand build tx. Gotta clean up some ltitle things and write another test or two.

### CODE REVIEW

File: test/v2/unit/coins/algo.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        spendableBalance: 650000000,
+        balanceString: '650000000',
+        confirmedBalanceString: '650000000',
+        spendableBalanceString: '650000000',
+        receiveAddress: {
+          id: '5a78de2bbfe424aa07aa131ec03c8dc1',
+          address: 'GBSGOXQNPULNPBX7XTHXP2GDBOZMHPIZOOA7CQXYTLHYKF5J2JDCF7LT',
+          chain: 0,
+          index: 0,
+          coin: 'txlm',
```
Comment (argjv) [inline test/v2/unit/coins/algo.ts]: talgo

Comment (twtaylorbitgo) [reply test/v2/unit/coins/algo.ts]: k. I changed all this. this mock wallet was copied from something else

File: src/v2/wallet.ts [hunk 2]
```diff
@@ -0,0 +1,1980 @@
+import common = require('../common');
```
Comment (tylerlevine) [inline src/v2/wallet.ts]: move to modules/core/src/v2

File: modules/core/package.json [hunk 3]
```diff
@@ -69,6 +70,9 @@
     "superagent-proxy": "^1.0.3"
   },
   "devDependencies": {
+    "@babel/cli": "^7.4.4",
```
Comment (tylerlevine) [inline modules/core/package.json]: Why are we adding babel?

File: modules/core/src/v2/wallet.ts [hunk 4]
```diff
@@ -1168,8 +1171,12 @@ Wallet.prototype.prebuildAndSignTransaction = function(params, callback) {
     // call prebuildTransaction and keychains-get in parallel
     // the prebuild can be overridden by providing an explicit tx
     const txPrebuildQuery = params.prebuildTx ? Promise.resolve(params.prebuildTx) : this.prebuildTransaction(params);
+
     const userKeychainQuery = this.baseCoin.keychains().get({ id: this._wallet.keys[0], reqId: params.reqId });
-    const [txPrebuild, userKeychain] = yield Promise.all([txPrebuildQuery, userKeychainQuery]);
+    const backupPublicKeys = this.baseCoin.keychains().get({ id: this._wallet.keys[1], reqId: params.reqId });
+    const bitgoPublicKey = this.baseCoin.keychains().get({ id: this._wallet.keys[2], reqId: params.reqId });
```
Comment (tylerlevine) [inline modules/core/src/v2/wallet.ts]: So we have to get the bitgo key for all coins now? :(

File: modules/core/src/v2/wallet.ts [hunk 5]
```diff
@@ -1180,7 +1187,16 @@ Wallet.prototype.prebuildAndSignTransaction = function(params, callback) {
...
+    // pass our three keys
+    const signingParams = _.extend({}, params, { 
+      txPrebuild: txPrebuild, 
+      wallet: { 
+        // this is the version of the multisig address at wallet creation time
+        addressVersion: this._wallet.coinSpecific.addressVersion
+      },
+      keychain: userKeychain, 
+      backupKeychain, 
+      bitgoKeychain });
```
Comment (tylerlevine) [inline modules/core/src/v2/wallet.ts]: Closing brace on next line

File: modules/core/test/v2/unit/coins/algo.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
+import { strict as assert } from 'assert';
```
Comment (tylerlevine) [inline modules/core/test/v2/unit/coins/algo.ts]: I don't think you're using this

File: modules/core/package.json (hunk 3 above)
Comment (twtaylorbitgo) [reply modules/core/package.json]: eh, algosdk won't build without it, we might need to re-factor algosdk-bitgo

File: modules/core/src/v2/wallet.ts (hunk 4 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/wallet.ts]: all pubs for algo

File: src/v2/wallet.ts (hunk 2 above)
Comment (twtaylorbitgo) [reply src/v2/wallet.ts]: yea, deleted this file completely, not sure what happened here, probably a bash stash

File: modules/core/package.json (hunk 3 above)
Comment (twtaylorbitgo) [reply modules/core/package.json]: FACTORED

File: modules/core/src/v2/wallet.ts (hunk 4 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/wallet.ts]: ok changed all this around

File: modules/core/test/v2/unit/coins/algo.ts (hunk 6 above)
Comment (twtaylorbitgo) [reply modules/core/test/v2/unit/coins/algo.ts]: removed

File: modules/core/src/v2/wallet.ts (hunk 5 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/wallet.ts]: yarp

File: modules/core/src/v2/coins/algo.ts [hunk 7]
```diff
@@ -1,35 +1,103 @@
...
-        return false;
+    if (!_.isNumber(addressVersion)) {
+      throw new Error('missing addressVersion parameter to sign transaction');
     }
+
+    const refinedTxData = txData;
+    refinedTxData.amount = parseInt(txData.amount, 10);
+
+    if (!_.has(txData, 'note')) {
+      refinedTxData.note = new Uint8Array(0);
```
[the remark this answers is not in the document]
Comment (twtaylorbitgo) [reply modules/core/src/v2/coins/algo.ts]: yeah... maybe I can put a comment? the problem is the sdk  will not take a null for note, it requires an empty array :(

File: modules/core/src/v2/coins/algo.ts [hunk 8]
```diff
@@ -1,35 +1,120 @@
 import BaseCoin = require('../baseCoin');
+import * as _ from 'lodash';
+
+const { MultiSigTransaction } = require('algosdk').Multisig;
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/algo.ts]: kind of funky to require algosdk twice (I know it's cached)

File: modules/core/src/v2/wallet.ts [hunk 9]
```diff
@@ -620,6 +620,16 @@ Wallet.prototype.addresses = function(params, callback) {
   .nodeify(callback);
 };
 
+/**
+ * Gets the keychains queries we will resolve
+ * @param reqId
+ * @returns {*}
+ */
+Wallet.prototype.keychainQueriesForSigning = function(reqId) {
```
Comment (tylerlevine) [inline modules/core/src/v2/wallet.ts]: I'd rather not expose this as part of the public api. This is trivial enough that it should probably just be inline in `signTransaction`

File: modules/core/test/v2/unit/coins/algo.ts [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+const nock = require('nock');
+nock.enableNetConnect();
+
+describe('ALGO:', function() {
+  let bitgo;
+  let basecoin;
+  let signedTxBase64;
+
+  before(function() {
+    bitgo = new TestV2BitGo({ env: 'test' });
```
Comment (tylerlevine) [inline modules/core/test/v2/unit/coins/algo.ts]: env: 'mock' in unit tests. This makes it much more obvious when you are missing a nock

Comment (twtaylorbitgo) [reply modules/core/test/v2/unit/coins/algo.ts]: k

File: modules/core/src/v2/wallet.ts (hunk 9 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/wallet.ts]: ah good point

File: modules/core/src/v2/coins/algo.ts (hunk 8 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/coins/algo.ts]: ha, ok changed

File: modules/core/package.json [hunk 11]
```diff
@@ -44,6 +44,7 @@
     "@types/mocha": "^5.2.6",
     "@types/nock": "^9.3.1",
     "@types/node": "^11.11.4",
+    "algosdk": "git+https://github.com/BitGo/algosdk-bitgo.git#0668c1412466beff2591a94c10c8bad06e21d872",
```
Comment (CapnMigraine) [inline modules/core/package.json]: I don't think you need to reference this particular commit, it's the latest one.

Comment (twtaylorbitgo) [reply modules/core/package.json]: hmm I don't know if this is matters, I can change it, it pins to what's in teh the yarn lock i'm pretty sure

Comment (CapnMigraine) [reply modules/core/package.json]: OK, if that's the case then lgtm

**Comment by twtaylorbitgo** [comment]
> no package-lock.json for core plz

my bad, removed

File: modules/core/package.json (hunk 11 above)
Comment (tylerlevine) [reply modules/core/package.json]: Don't think we need to pin a hash here - that's what the yarn.lock is for

File: modules/core/src/v2/coins/algo.ts [hunk 12]
```diff
@@ -1,11 +1,15 @@
 import BaseCoin = require('../baseCoin');
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/algo.ts]: Can you configure prettier on this file?

You do that by adding this at the top of the file:
```
/*
 * @prettier 
 */
```

Then your code gets auto formatted upon commit.

File: modules/core/src/v2/wallet.ts [hunk 13]
```diff
@@ -1168,8 +1171,13 @@ Wallet.prototype.prebuildAndSignTransaction = function(params, callback) {
     // call prebuildTransaction and keychains-get in parallel
     // the prebuild can be overridden by providing an explicit tx
     const txPrebuildQuery = params.prebuildTx ? Promise.resolve(params.prebuildTx) : this.prebuildTransaction(params);
-    const userKeychainQuery = this.baseCoin.keychains().get({ id: this._wallet.keys[0], reqId: params.reqId });
-    const [txPrebuild, userKeychain] = yield Promise.all([txPrebuildQuery, userKeychainQuery]);
+
+    const self = this;
+    const keychains = yield Promise.all(function keychainQueriesForSigning() {
```
Comment (tylerlevine) [inline modules/core/src/v2/wallet.ts]: Let's pull this out above the call to promise.all. I don't think the IIFE is needed here

File: modules/core/test/v2/unit/coins/algo.ts [hunk 14]
```diff
@@ -1,18 +1,32 @@
+import * as Promise from 'bluebird';
 import * as crypto from 'crypto';
 import * as algosdk from 'algosdk';
 
+const co = Promise.coroutine;
+const Wallet = require('../../../../src/v2/wallet');
+
+import 'should';
```
Comment (tylerlevine) [inline modules/core/test/v2/unit/coins/algo.ts]: move to top.

File: modules/core/test/v2/unit/coins/algo.ts [hunk 15]
```diff
@@ -1,18 +1,32 @@
...
+
 const TestV2BitGo = require('../../../lib/test_bitgo');
 
+const nock = require('nock');
+nock.enableNetConnect();
```
Comment (tylerlevine) [inline modules/core/test/v2/unit/coins/algo.ts]: disableNetConnect - this is a unit test

File: modules/core/test/v2/unit/coins/algo.ts [hunk 16]
```diff
@@ -64,4 +78,105 @@ describe('ALGO:', function() {
...
+
+    const bitgoKeychain = {
+      pub: '3GFMZSPBX3XQL22IBAUF2UQFAZCOF5QFFVRRLEHFPHFLN5NY7WQJSSZRGQ',
+      // normally this wouldn't be here but we include it for signing purposes
+      prv: 'DLHE27PAI335VEQUNX7R2442SSJZUYISBL7OOR3V34AYNTH7YWI6XMGVVE'
+    }
+
+    before(co(function *() {
+      basecoin = bitgo.coin('talgo');
+      walletData = {
```
Comment (tylerlevine) [inline modules/core/test/v2/unit/coins/algo.ts]: Consider moving some of this to fixtures/

File: modules/core/test/v2/unit/coins/algo.ts (hunk 15 above)
Comment (twtaylorbitgo) [reply modules/core/test/v2/unit/coins/algo.ts]: ack

File: modules/core/test/v2/unit/coins/algo.ts (hunk 16 above)
Comment (twtaylorbitgo) [reply modules/core/test/v2/unit/coins/algo.ts]: k

File: modules/core/test/v2/unit/coins/algo.ts (hunk 14 above)
Comment (twtaylorbitgo) [reply modules/core/test/v2/unit/coins/algo.ts]: k

File: modules/core/src/v2/wallet.ts (hunk 13 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/wallet.ts]: ok

File: modules/core/src/v2/coins/algo.ts (hunk 12 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/coins/algo.ts]: sure

**Pull request merged**